packages/edit-site/src/store/actions.js

```diff
@@ -107,19 +107,39 @@ export function setTemplatePart( templatePartId ) {
 /**
  * Resolves the template for a page and sets them.
  *
- * @param {Object} page         The page object.
- * @param {string} page.type    The page type.
- * @param {string} page.slug    The page slug.
- * @param {string} page.path    The page path.
- * @param {Object} page.context The page context.
+ * @param {Object}  page         The page object.
+ * @param {string}  page.type    The page type.
+ * @param {string}  page.slug    The page slug.
+ * @param {string}  page.path    The page path.
+ * @param {Object}  page.context The page context.
+ * @param {number?} templateId   An optional template ID for the page to avoid a second fetch if we already know it.
  *
  * @return {Object} Action object.
  */
-export function* setPage( page ) {
-	const templateId = yield findTemplate( page.path );
+export function* setPage( page, templateId ) {
+	const resolvedTemplate = templateId ?? ( yield findTemplate( page.path ) );
 	return {
 		type: 'SET_PAGE',
 		page,
-		templateId,
+		templateId: resolvedTemplate,
 	};
 }
+
+/**
+ * Sets up the initial template and page for edit site, and fetches other
+ * information it needs.
+ *
+ * @param {Object} initialPage The initial page to load in the site editor.
+ */
+export function* setupState( initialPage ) {
```

Comment:
This doesn't feel like a great place to put this logic. Where else could it go?

Reply:
TBH I'm generally not a big fan of actions that set initial state. I prefer reducers with a reasonable initial state; and, if we need to fetch information for a more meaningful de-facto initial state, I'd prefer an action that also makes sense to set state at a later time.

Reply:
I can maybe rephrase to -- what does `setupState` give us that `setPage` doesn't?

Reply:
it fetches and stores the homepage template, which is needed to display which "template" is the homepage. this can be different from the initial page if we ever want to change which page is loaded at start

I agree with you in general, and wanted to block initializing the editor on async data load (which is why this is not part of the initial state).

It sounds like we just want to run a couple of extra actions when initializing the editor, rather than calling a setup action?

Reply:
> it fetches and stores the homepage template, which is needed to display which "template" is the homepage. this can be different from the initial page if we ever want to change which page is loaded at start

I'll play the YAGNI card here -- let's not prepare for that case before it arrives :slightly_smiling_face: 

> I agree with you in general, and wanted to block initializing the editor on async data load (which is why this is not part of the initial state).
> 
> It sounds like we just want to run a couple of extra actions when initializing the editor, rather than calling a setup action?

I think I'd overall prefer that architecturally, yeah :thinking: